lib/internal/source_map.js

```diff
@@ -0,0 +1,122 @@
+'use strict';
+
+// See https://sourcemaps.info/spec.html for SourceMap V3 specification.
+const { Buffer } = require('buffer');
+const debug = require('internal/util/debuglog').debuglog('source_map');
+const { dirname, resolve } = require('path');
+const fs = require('fs');
+const {
+  normalizeReferrerURL,
+} = require('internal/modules/cjs/helpers');
+const { JSON, Object } = primordials;
+const sourceMapCache = new Map();
```

Comment:
Some modules and applications interact with the require cache and load/unload modules at runtime. This new Map will leak memory in that case.

Reply:
Good catch, can be a `WeakMap`

Reply:
@mcollina is this a problem for `internal` modules, I don't believe they show up in `require.cache` object at least.

I'm not used to using the `WeakMap` data type, but it doesn't seem like it works well for a cache, as ideally I need to reference the `source-map` cache by a string key.

Reply:
@bcoe are the source maps the same even if the modules replace themselves in the cache? ie. if I do `require.cache['foo'] = require.cache['bar']` what is the expected value of the source map? does it adopt `bar`'s? or keep the original `foo`'s? Likely attaching them to the `Module` objects is the most sensible.

Reply:
@bmeck, this was @joyeecheung's suggestion as well, this seem quite reasonable to me; I'll work on a patch that does so.

:+1: @benjamingr @mcollina I feeling like I know how to proceed.

Reply:
@bcoe faster and less GC burden would be if this could be done via private fields. Since `Module` isn't an ES6 class however, it would need to be done with the super override trick. IDK if that matters here.

Reply:
@joyeecheung @bmeck @mcollina @benjamingr see:

https://github.com/nodejs/node/pull/28960/commits/51943f4847bbd38516a9d026f8d759599ba5da46

I've introduced a `WeakMap` for modules loaded using "cjs/loader.js"; correct me if I'm wrong, but I believe we can continue using a `Map` for caching modules loaded via ESM, because we don't expose a similar mechanism to users for deleting cached objects?

Reply:
> @mcollina is this a problem for internal modules, I don't believe they show up in require.cache object at least.

@bcoe I don't understand this statement.

As @bmeck states, attaching this to the module object would be better.